### PR TITLE
[SofaMatrix] A new plugin adding tools for linear system matrix

### PR DESCRIPTION
**Pull request by alxbilger**
A long PR but it is actually divided in 3 parts. A link is provided for each part pointing to a comparison with the previous part.

# 1. Introduction of the plugin

A new plugin is introduced to welcome new components. It is enabled by default.
https://github.com/sofa-framework/sofa/compare/master...alxbilger:sofamatrix?expand=1

The plugin may deserve a better name.

# 2. Move CImgPlugin features on matrices

I introduced features related to matrices in CImgPlugin in a previous PR. It did not make sense. It is preferable to have them in a higher-level plugin, so I moved them in SofaMatrix.
https://github.com/alxbilger/sofa/compare/sofamatrix...sofamatrix_cimg

# 3. Introduction of a component to visualize the matrix

A component is introduced. It gets the matrix from the linear system and convert it to an bitmap to visualize it directly in the GUI.
https://github.com/alxbilger/sofa/compare/sofamatrix_cimg...sofamatrix_gui

image

# To do:

- [x] Expand the bitmap in height
- [ ] Update the bitmap when the image changes during the simulation
- [ ] Do we move `GlobalSystemMatrixExporter` into SofaMatrix?

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: applications/plugins/SofaMatrix/src/SofaMatrix/SimpleImageViewerWidget.h [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+{
+public:
+    SimpleImageViewerGraphWidget(QWidget * parent, const SimpleBitmapType& plane)
+    : BaseSimpleImageViewerGraphWidget(parent)
+    , m_imageplane(&plane)
+    {
+        readFromData(plane);
+        const QSizePolicy sizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+        setSizePolicy(sizePolicy);
+
```
Comment (epernod) [inline applications/plugins/SofaMatrix/src/SofaMatrix/SimpleImageViewerWidget.h]:
```suggestion
        QVBoxLayout* vbox = new QVBoxLayout();
        vbox->addStretch();
        this->setLayout(vbox);
        
```

Comment (epernod) [reply applications/plugins/SofaMatrix/src/SofaMatrix/SimpleImageViewerWidget.h]: Tada! qt layout logic magic!

**Review by epernod — APPROVED** [review]

**Comment by epernod** [comment]
image

File: applications/plugins/SofaMatrix/src/SofaMatrix/SimpleImageViewerWidget.h (hunk 1 above)
Comment (alxbilger) [reply applications/plugins/SofaMatrix/src/SofaMatrix/SimpleImageViewerWidget.h]: Thanks a lot! You're my Qt expert now! Thanks again

File: applications/plugins/SofaMatrix/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    )
+    list(APPEND SOFA_MODULES
+        SofaGuiQt
+    )
+
+    find_package(Qt5 COMPONENTS Core Gui Widgets QUIET)
+    if(Qt5Core_FOUND)
+        qt5_wrap_cpp(MOC_FILES ${MOC_HEADER_FILES})
+    else()
+        find_package(Qt4 COMPONENTS qtcore)
```
Comment (fredroy) [inline applications/plugins/SofaMatrix/CMakeLists.txt]: ... you want to support qt4 ? 😵
I would advise to focus more on qt6 🧐

Comment (alxbilger) [reply applications/plugins/SofaMatrix/CMakeLists.txt]: you are right. It is copy/pasted from the `image` plugin.

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Review by fredroy — APPROVED** [review]

**Pull request merged**